presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -631,13 +631,13 @@ private boolean isRecoverable(List<ExecutionFailureInfo> failures)
 
     private synchronized void updateFinalTaskInfo(TaskInfo finalTaskInfo)
     {
-        tasksWithFinalInfo.add(finalTaskInfo.getTaskId());
+        tasksWaitingForFinalInfo.remove(finalTaskInfo.getTaskId());
         checkAllTaskFinal();
     }
 
     private synchronized void checkAllTaskFinal()
     {
-        if (stateMachine.getState().isDone() && tasksWithFinalInfo.containsAll(allTasks)) {
```

Comment:
So is tasksWithFinalInfo.containsAll(allTasks) the slow method ? With 1000's of tasks per stage maybe this is the thing that is being slow and is this the entire source of lock contention ?

Reply:
Also if this is indeed the only real change, I am also not understanding why it is especially slow esp in the beginning when tasksWithFinalInfo is small (it starts out empty), and then is added one by one as tasks finish. This method is also O(tasksWithFinalInfo.size()).

Reply:
The JFR data shows almost all the wait are on this method, and thread dump shows sometime one of this call holds the lock. We just made our best guess, optimize and profile gain. 

I think it is combination of this method is relatively slow, and more importantly it is called too many times. In one example, we observed combine 20 mins waiting time on one lock in 60s profiling section.